### PR TITLE
Fix test failure on same trigger

### LINKED ISSUE #1347: Two coroutines that fail on the same trigger are not handled gracefully by the scheduler.
```python
@cocotb.coroutine
async def coroutine1(timer):
    await timer
    assert False

@cocotb.coroutine
async def coroutine2(timer):
    await timer
    assert False

@cocotb.test()
async def test_top(dut):
    timer = triggers.Timer(100, units='ns')                                                                                            
    cocotb.fork(coroutine1(timer))                                                                                                     
    cocotb.fork(coroutine2(timer))
    await timer
```
This will fail when `_force_outcome` is called by the scheduler for the second time.  The scheduler needs to keep track of whether any coroutines that have been scheduled by this trigger have already failed.

### PR DESCRIPTION
**Pull request by ktbarrett**
Closes #1347.

### CODE REVIEW

**Comment by imphil** [comment]
Thanks @ktbarrett for looking into this long-standing issue!

@eric-wieser would you mind having a look?

Since 1.7.0 has settled nicely I'm going the cautious route and not picking this up for 1.7.0 (scheduler changes are too much risk for me at this point). I'm happy, however, to do a 1.7.1 in two weeks or so with this change included.

File: cocotb/scheduler.py [hunk 1]
```diff
@@ -1072,6 +1074,12 @@ def _cleanup(self):
                     self.log.debug("Killing %s" % str(coro))
                 coro.kill()
 
+        # if there are coroutines being scheduled when the test ends, kill them (gh-1347)
```
Comment (eric-wieser) [inline cocotb/scheduler.py:1077]: It's not immediately clear to me that `_cleanup` should be allowed to run while in the middle of the `_event_loop`.

Comment (ktbarrett) [reply cocotb/scheduler.py:1077]: Intuitively, you are right. Unfortunately, it doesn't work that way. Maybe one day, after a significant refactor, the scheduler will be a little more intuitive. I've already tried a few times to do this without a lot of success.

If a task fails it is `_unschedule`d. If there are no waiters on it, then it `_abort_test`s. The test is then `_unschedule`d which results in a call to `_cleanup`.

Comment (cmarqu) [reply cocotb/scheduler.py:1077]: At some point we should probably have some PlantUML state diagrams or something like that in the docs.

**Review by marlonjames — APPROVED** [review]
This is the most straight-forward fix given the current scheduler design.

**Pull request merged**